### PR TITLE
Update `pkg/controller/cronjob/` for contextual logging

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Use contextual logging for `pkg/controller/cronjob/`

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note.
Migrated the CronJob controller (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

### CODE REVIEW

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 1]
```diff
@@ -196,10 +200,10 @@ func (jm *ControllerV2) sync(ctx context.Context, cronJobKey string) (*time.Dura
 
 	cronJobCopy, requeueAfter, updateStatus, err := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
 	if err != nil {
-		klog.V(2).InfoS("Error reconciling cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "err", err)
+		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:203]: /lgtm cancel

```suggestion
		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)
```

Please also change it below.

Comment (mengjiao-liu) [reply pkg/controller/cronjob/cronjob_controllerv2.go]: Oh, I didn't notice.
Updated.

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 2]
```diff
@@ -196,10 +200,10 @@ func (jm *ControllerV2) sync(ctx context.Context, cronJobKey string) (*time.Dura
 
 	cronJobCopy, requeueAfter, updateStatus, err := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
 	if err != nil {
-		klog.V(2).InfoS("Error reconciling cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "err", err)
+		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)
 		if updateStatus {
 			if _, err := jm.cronJobControl.UpdateStatus(ctx, cronJobCopy); err != nil {
-				klog.V(2).InfoS("Unable to update status for cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "resourceVersion", cronJob.ResourceVersion, "err", err)
+				logger.V(2).Info("Unable to update status for cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "resourceVersion", cronJob.ResourceVersion, "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:206]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 3]
```diff
@@ -213,13 +217,13 @@ func (jm *ControllerV2) sync(ctx context.Context, cronJobKey string) (*time.Dura
 	// Update the CronJob if needed
 	if updateStatus {
 		if _, err := jm.cronJobControl.UpdateStatus(ctx, cronJobCopy); err != nil {
-			klog.V(2).InfoS("Unable to update status for cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "resourceVersion", cronJob.ResourceVersion, "err", err)
+			logger.V(2).Info("Unable to update status for cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "resourceVersion", cronJob.ResourceVersion, "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:220]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 4]
```diff
@@ -213,13 +217,13 @@ func (jm *ControllerV2) sync(ctx context.Context, cronJobKey string) (*time.Dura
...
 			return nil, err
 		}
 	}
 
 	if requeueAfter != nil {
-		klog.V(4).InfoS("Re-queuing cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "requeueAfter", requeueAfter)
+		logger.V(4).Info("Re-queuing cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "requeueAfter", requeueAfter)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:226]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 5]
```diff
@@ -393,7 +397,7 @@ func (jm *ControllerV2) updateCronJob(old interface{}, curr interface{}) {
 		if err != nil {
 			// this is likely a user error in defining the spec value
 			// we should log the error and not reconcile this cronjob until an update to spec
-			klog.V(2).InfoS("Unparseable schedule for cronjob", "cronjob", klog.KRef(newCJ.GetNamespace(), newCJ.GetName()), "schedule", newCJ.Spec.Schedule, "err", err)
+			logger.V(2).Info("Unparseable schedule for cronjob", "cronjob", klog.KRef(newCJ.GetNamespace(), newCJ.GetName()), "schedule", newCJ.Spec.Schedule, "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:400]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 6]
```diff
@@ -494,34 +498,35 @@ func (jm *ControllerV2) syncCronJob(
 		return cronJob, nil, updateStatus, nil
 	}
 
+	logger := klog.FromContext(ctx)
 	if timeZoneEnabled && cronJob.Spec.TimeZone != nil {
 		if _, err := time.LoadLocation(*cronJob.Spec.TimeZone); err != nil {
 			timeZone := pointer.StringDeref(cronJob.Spec.TimeZone, "")
-			klog.V(4).InfoS("Not starting job because timeZone is invalid", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "timeZone", timeZone, "err", err)
+			logger.V(4).Info("Not starting job because timeZone is invalid", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "timeZone", timeZone, "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:505]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 7]
```diff
@@ -494,34 +498,35 @@ func (jm *ControllerV2) syncCronJob(
...
 			jm.recorder.Eventf(cronJob, corev1.EventTypeWarning, "UnknownTimeZone", "invalid timeZone: %q: %s", timeZone, err)
 			return cronJob, nil, updateStatus, nil
 		}
 	}
 
 	if cronJob.Spec.Suspend != nil && *cronJob.Spec.Suspend {
-		klog.V(4).InfoS("Not starting job because the cron is suspended", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()))
+		logger.V(4).Info("Not starting job because the cron is suspended", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()))
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:512]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 8]
```diff
@@ -494,34 +498,35 @@ func (jm *ControllerV2) syncCronJob(
...
 		return cronJob, nil, updateStatus, nil
 	}
 
 	sched, err := cron.ParseStandard(formatSchedule(timeZoneEnabled, cronJob, jm.recorder))
 	if err != nil {
 		// this is likely a user error in defining the spec value
 		// we should log the error and not reconcile this cronjob until an update to spec
-		klog.V(2).InfoS("Unparseable schedule", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "schedule", cronJob.Spec.Schedule, "err", err)
+		logger.V(2).Info("Unparseable schedule", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "schedule", cronJob.Spec.Schedule, "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:520]: KRef -> KObj

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 9]
```diff
@@ -494,34 +498,35 @@ func (jm *ControllerV2) syncCronJob(
...
 		return cronJob, nil, updateStatus, nil
 	}
 
-	scheduledTime, err := getNextScheduleTime(*cronJob, now, sched, jm.recorder)
+	scheduledTime, err := getNextScheduleTime(logger, *cronJob, now, sched, jm.recorder)
 	if err != nil {
 		// this is likely a user error in defining the spec value
 		// we should log the error and not reconcile this cronjob until an update to spec
-		klog.V(2).InfoS("invalid schedule", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "schedule", cronJob.Spec.Schedule, "err", err)
+		logger.V(2).Info("invalid schedule", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "schedule", cronJob.Spec.Schedule, "err", err)
```
Comment (pohly) [inline pkg/controller/cronjob/cronjob_controllerv2.go:529]: KRef -> KObj

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 10]
```diff
@@ -572,21 +577,20 @@ func (jm *ControllerV2) syncCronJob(
...
+		logger.V(4).Info("Not starting job because prior execution is still running and concurrency policy is Forbid", "cronjob", klog.KObj(cronJob))
 		jm.recorder.Eventf(cronJob, corev1.EventTypeNormal, "JobAlreadyActive", "Not starting job because prior execution is running and concurrency policy is Forbid")
 		t := nextScheduledTimeDuration(*cronJob, sched, now)
 		return cronJob, t, updateStatus, nil
 	}
 	if cronJob.Spec.ConcurrencyPolicy == batchv1.ReplaceConcurrent {
 		for _, j := range cronJob.Status.Active {
-			klog.V(4).InfoS("Deleting job that was still running at next scheduled start time", "job", klog.KRef(j.Namespace, j.Name))
-
+			logger.V(4).Info("Deleting job that was still running at next scheduled start time", "job", klog.KRef(j.Namespace, j.Name))
```
Comment (mengjiao-liu) [inline pkg/controller/cronjob/cronjob_controllerv2.go:587]: `klog.KObj`  is not used here for the following reason:
```
Cannot use 'j' (type corev1.ObjectReference) as the type KMetadata Type does not implement 'KMetadata' as some methods are missing: GetName() string GetNamespace() string
```

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 11]
```diff
@@ -540,7 +545,7 @@ func (jm *ControllerV2) syncCronJob(
 		tooLate = scheduledTime.Add(time.Second * time.Duration(*cronJob.Spec.StartingDeadlineSeconds)).Before(now)
 	}
 	if tooLate {
-		klog.V(4).InfoS("Missed starting window", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()))
+		logger.V(4).Info("Missed starting window", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()))
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/controller/cronjob/cronjob_controllerv2.go]: ok. I have searched for such calls and replaced all of them. Thanks!

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by pohly** [comment]
/assign @soltysh

For approval.

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 12]
```diff
@@ -746,13 +750,13 @@ func isJobInActiveList(job *batchv1.Job, activeJobs []corev1.ObjectReference) bo
 }
 
 // deleteJob reaps a job, deleting the job, the pods and the reference in the active list
-func deleteJob(cj *batchv1.CronJob, job *batchv1.Job, jc jobControlInterface, recorder record.EventRecorder) bool {
-	nameForLog := fmt.Sprintf("%s/%s", cj.Namespace, cj.Name)
+func deleteJob(ctx context.Context, cj *batchv1.CronJob, job *batchv1.Job, jc jobControlInterface, recorder record.EventRecorder) bool {
+	logger := klog.FromContext(ctx)
```
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/controller/cronjob/cronjob_controllerv2.go]: Indeed, it is better to pass the logger instance instead.

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 13]
```diff
@@ -691,33 +695,31 @@ func (jm *ControllerV2) getFinishedStatus(j *batchv1.Job) (bool, batchv1.JobCond
...
 // deleteJob reaps a job, deleting the job, the pods and the reference in the active list
-func deleteJob(cj *batchv1.CronJob, job *batchv1.Job, jc jobControlInterface, recorder record.EventRecorder) bool {
-	nameForLog := fmt.Sprintf("%s/%s", cj.Namespace, cj.Name)
-
+func deleteJob(logger klog.Logger, cj *batchv1.CronJob, job *batchv1.Job, jc jobControlInterface, recorder record.EventRecorder) bool {
 	// delete the job itself...
 	if err := jc.DeleteJob(job.Namespace, job.Name); err != nil {
 		recorder.Eventf(cj, corev1.EventTypeWarning, "FailedDelete", "Deleted job: %v", err)
-		klog.Errorf("Error deleting job %s from %s: %v", job.Name, nameForLog, err)
+		logger.Error(err, "Error deleting job", "job", klog.KObj(job), "cronjob", klog.KObj(cj))
```
Comment (atiratree) [inline pkg/controller/cronjob/cronjob_controllerv2.go:722]: nit:

```suggestion
		logger.Error(err, "Error deleting job from cronjob", "job", klog.KObj(job), "cronjob", klog.KObj(cj))
```

**Review by atiratree — COMMENTED** [review]
/lgtm

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 13 above)
Comment (mengjiao-liu) [reply pkg/controller/cronjob/cronjob_controllerv2.go]: Thanks! Updated.

**Comment by pohly** [comment]
/assign @soltysh 

For approval.

**Review by soltysh — APPROVED** [review]
/approve

**Comment by liggitt** [comment]
as soon as this merged, the run time of the TestCronJobLaunchesPodAndCleansUp test increased from ~30 seconds to ~90 seconds... https://testgrid.k8s.io/sig-release-master-blocking#integration-master&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=TestCronJobLaunchesPodAndCleansUp

is this change messing with the cancel behavior of contexts used to stop controllers?

**Comment by pohly** [comment]
Issue: https://github.com/kubernetes/kubernetes/issues/116378
Fix: https://github.com/kubernetes/kubernetes/pull/116381

**Pull request merged**